File: gpt-3.5-ft-large/stage2/acc_memcpy_from_device.c
#include "acc_testsuite.h"
#ifndef T1
//T1: acc_memcpy_from_device,V:2.7-3.3
int test1(){
    int err = 0;
    srand(SEED);
    real_t *a = (real_t *)malloc(n * sizeof(real_t));
    real_t *b = (real_t *)malloc(n * sizeof(real_t));
    real_t *c = (real_t *)malloc(n * sizeof(real_t));
    real_t *hostdata = (real_t *)malloc(3 * n * sizeof(real_t));
    real_t *devdata;

    for (int x = 0; x < n; ++x){
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = rand() / (real_t)(RAND_MAX / 10);
        c[x] = 1;
    }

    devdata = acc_deviceptr(hostdata);

    #pragma acc enter data copyin(a[0:n], b[0:n]) create(hostdata[0:3*n])
    #pragma acc parallel present(hostdata[0:3*n], devdata[0:3*n])
    {
        #pragma acc loop
        for (int x = 0; x < n; ++x){
            devdata[x] = a[x];
            devdata[n + x] = b[x];
        }
        #pragma acc loop
        for (int x = 0; x < 3 * n; ++x){
                hostdata[x] = devdata[x];
        }
    }
    for (int x = 0; x < n; ++x){
        if (fabs(hostdata[x] - a[x]) > PRECISION){
            err += 1;
        }
        if (fabs(hostdata[n + x] - b[x]) > PRECISION){
            err += 1;
        }
    }

    #pragma acc enter data copyin(hostdata[0:3*n]) copyout(c[0:n])
    #pragma acc parallel present(c[0:n], hostdata[0:3*n], devdata[0:3*n])
    {
        #pragma acc loop
        for (int x = 0; x < n; ++x){
            c[x] += devdata[x] + devdata[n + x];
        }
    }

    for (int x = 0; x < n; ++x){
        if (fabs(c[x] - (1 + a[x] + b[x])) > PRECISION){
            err += 1;
        }
    }

    #pragma acc exit data copyout(a[0:n], b[0:n], c[0:n]) delete(hostdata[0:3*n])
    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
    return failcode;
}

File: gpt-3.5-ft-large/stage2/acc_memcpy_from_device.f90
#ifndef T1
!T1:acc memcpy from device,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0
        REAL(8),DIMENSION(LOOPCOUNT):: a, b
        INTEGER ,DIMENSION(1) :: devtest
        INTEGER :: multi
        multi = 0


        devtest(1) = 1
        #        ifdef _OPENACC
!$acc enter data copyin(devtest(1:1))
#        endif



        SEEDDIM(1) = 1
#        ifdef _OPENACC
!$acc data copyout(a(1:LOOPCOUNT), b(1:LOOPCOUNT))
#        Endif
        CALL RANDOM_SEED()
        CALL RANDOM_NUMBER(a)
        b = 0
        IF (acc_is_present(a(1:LOOPCOUNT)) .eqv. 1) THEN
          a = 1.0
        END IF
        IF (acc_is_present(a(1:LOOPCOUNT)) .eqv. 0) THEN
          a = 2.0
        END IF
        #        ifdef _OPENACC
!$acc exit data copyout(a(1:LOOPCOUNT))
#        Endif
        

        CALL acc_memcpy_from_device(b(1), a(1), LOOPCOUNT*4)

        DO WHILE (MULTI < 10)
          CALL acc_memcpy_from_device(b(1), a(1), LOOPCOUNT * 4)
          DO i = 1, LOOPCOUNT
            IF (abs(b(i) - a(i)) .gt. PRECISION) THEN
              errors = errors + 1
            END IF
          END DO
          multi = multi + 1
        END DO



        DELETEDEVCOLL
        IF ((FOR_TESTING_HOSTNEEDEDMODES .eq. 1) .or. (acc_get_device_type() .eq. acc_device_altera))
          THEN
          IF (errors .eq. 0) THEN
            test1 = .FALSE.
          ELSE
            test1 = .TRUE.
          END IF
        ELSE
          test1 = .FALSE.
        END IF
      END
#endif

        DO WHILE (MULTI < 10)
          CALL acc_memcpy_from_device(b(1), a(1), LOOPCOUNT * 4)
          DO i = 1, LOOPCOUNT
            IF (abs(b(i) - a(i)) .gt. PRECISION) THEN
              errors = errors + 1
            END IF
          END DO
          multi = multi + 1
        END DO



        DELETEDEVCOLL
        IF ((FOR_TESTING_HOSTNEEDEDMODES .eq. 1) .or. (acc_get_device_type() .eq. acc_device_altera))
          THEN
          IF (errors .eq. 0) THEN
            test1 = .FALSE.
          ELSE
            test1 = .TRUE.
          END IF
        ELSE
          test1 = .FALSE.
        END IF
      END
#endif
      


      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
#ifndef T1
        LOGICAL :: test1
#endif
        failed = .FALSE.
        failcode = 0
#ifndef T1
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
#endif
        CALL EXIT (failcode)
      END PROGRAM